Android: Add the lkft-android project as reference for the Android jobs support

liuyq opened the pull request:
This change adds the support for the android jobs, and added
the lkft-android project with the x15 device and the boot testcase
added as reference.

It needs specify the --testplan-path to projects/lkft-android,
and the --testplan-device-path to projects/lkft-android/devices,
for the submit_for_testing.py script to be run like this:
```sh
$ ./submit_for_testing.py \
         --variables ${PWD}/projects/lkft-android/variables.ini \
         --testplan-device-path ${PWD}/projects/lkft-android/devices/ \
         --testplan-path ${PWD}/projects/lkft-android/  \
         --device-type x15 \
         --test-case boot.yaml \
         --dry-run
```

Signed-off-by: Yongqin Liu <yongqin.liu@linaro.org>

liuyq commented:
@roxell please help check if there is any problem to have this change merged.

liuyq commented on lava_test_plans/projects/lkft-android/fastboot.jinja2:
```diff
@@ -16,7 +16,13 @@
+  group:
+    - {{ TEST_LAVA_JOB_GROUP }}
+{% endblock visibility %}
+
```
could you please update the upstream side with this feature, then I will update here to use that feature.

roxell replied:
https://github.com/Linaro/lava-test-plans/pull/226

liuyq replied:
thanks, I just dropped this commit from this PR, and will submit it again after #226 merged.

liuyq commented on lava_test_plans/projects/lkft-android/boot-empty.yaml:
```diff
@@ -0,0 +1 @@
+../../testcases/boot.yaml
```
why you want to run like the following:
```
python3 ./submit_for_testing.py
--device-type x15
--variables android-variables.ini
--test-case boot.yaml
--dry-run
````
from what I understand, this command is to use the generic x15 device settings and the generic boot testcase to generate the lava definition, but with android variables.
any meaning to generate the generic definition with android variables?

liuyq replied:
with the following command, if there is no boot-empty.yaml link file, I guess it will use the boot.yaml under lava_test_plans/projects/lkft-android, which is the link of ../../testcases/android-boot.yaml.
```
python3 ./submit_for_testing.py
--device-type x15
--variables ./lava_test_plans/projects/lkft-android/variables.ini
--testplan-path ./lava_test_plans/projects/lkft-android/
--testplan-device-path ./lava_test_plans/projects/lkft-android/devices/
--test-case boot.yaml
--dry-run
```

liuyq replied:
the link of the link of ../../testcases/android-boot.yaml. will be added later

liuyq replied:
hmm, I think you are right, with this change, the lava_test_plans/projects/lkft-android/boot-empty.yaml  link could be removed, the testcases/boot.yaml will be used, and after we added the link to ../../testcases/android-boot.yaml,
if necessary we could add the boot-empty.yaml link again.

liuyq commented on lava_test_plans/include/test_target.jinja2:
```diff
@@ -1,5 +1,11 @@
 - test:
+{% if use_docker_image_for_test_target == true %}
+    docker:
+      image: {{ DOCKER_IMAGE }}
```
will update the name to DOCKER_IMAGE_FOR_TEST here

liuyq commented on lava_test_plans/projects/lkft-android/fastboot.jinja2:
```diff
@@ -23,6 +23,11 @@
+secrets:
+  ARTIFACTORIAL_TOKEN: "{{ ARTIFACTORIAL_TOKEN }}"
+{% endblock settings %}
+
```
that case the fastboot.jinja2 should not be used.

roxell replied:
wait with this patch until you add tests that uses the ARTIFACTORIAL_(URL|TOKEN) and we will see more clearly where it fits in. ok?

liuyq replied:
sure, I dropped this commit for this PR now.

liuyq commented on lava_test_plans/include/fastboot.jinja2:
```diff
@@ -139,9 +154,93 @@ reboot_to_fastboot: {{ reboot_to_fastboot }}
+{% endif %}
+{% if partition_boot_b == true %}
+      boot_b:
+        url: {{ BOOT_URL }}
```
updated

liuyq commented on lava_test_plans/include/fastboot.jinja2:
```diff
@@ -139,9 +154,93 @@ reboot_to_fastboot: {{ reboot_to_fastboot }}
+{% endif %}
+{% if partition_boot_a == true %}
+      boot_a:
+        url: {{ BOOT_URL }}
```
updated to use the format of BOOT_A_URL, and the same is done for the VBMETA_A and VBMETA_B partitions

roxell commented on lava_test_plans/include/test_target.jinja2:
```diff
@@ -1,5 +1,11 @@
+
 {% if enable_tests is defined and enable_tests %}
 - test:
+{% if use_docker_image_for_test_target == true %}
```
If this is used in the variable.(ini|yaml) file it should be with uppercase

liuyq replied:
thanks, will update to use uppercase

liuyq commented on lava_test_plans/include/fastboot.jinja2:
```diff
@@ -19,6 +19,7 @@
 {% set auto_login = auto_login|default(true) %}
 {% set boot_method = boot_method|default("fastboot") %}
 {% set fastboot_deploy_timeout = fastboot_deploy_timeout|default(40) %}
+{% set download_postprocess_required = download_postprocess_required|default(true) %}
```
lkft-android jobs need to have the postprocess to recreate the android images.

roxell replied:
so if lkft-android also needs it why do we want to introduce a new variable then?

Can't this patch be dropped until a project starts to see the benefit of not using the postprocess step?

liuyq replied:
lkft-android needs to use  postprocess, it also has the case that does not need to use postprocess, that's why this variable is added. 
if we do not add this variable, how do we make the postprocess not generated in the lava job file like for the case here: https://lkft.validation.linaro.org/scheduler/job/5983178/definition

liuyq commented on lava_test_plans/include/fastboot.jinja2:
```diff
@@ -140,6 +154,90 @@ reboot_to_fastboot: {{ reboot_to_fastboot }}
+        url: {{ BOOT_B_URL }}
+{% if BOOT_B_URL_COMP is defined %}
+        compression: {{ BOOT_B_URL_COMP }}
+{% endif %}
```
why? BOOT_LABEL is for the boot partition, and here boot_a and boot_a are separate partitions from it.

roxell replied:
something like this http://ix.io/4i80

liuyq replied:
thanks, updated to use the boot partition settings for the boot_a partition.

liuyq commented:
> if you can add some tests would be great too

no plan here, let's have something merged first.

liuyq commented on lava_test_plans/include/fastboot.jinja2:
```diff
@@ -140,6 +154,90 @@ reboot_to_fastboot: {{ reboot_to_fastboot }}
+{% endif %}
+{% endif %}
+{% if partition_userdata == true %}
+      userdata:
```
rootfs is one logic concept, while userdata is a real partition, I think it's the place where rootfs is used to set the correct partition it will use, instead of setting here to use the rootfs_label. If for some devices, the system or super partition is used as the rootfs for the lts setup, should the rootfs label be changed to system or super?

liuyq commented:
> if you can add some tests would be great too

ok, I updated the test/test_lava_test_plans.py file with the lkft-android project, but I am not sure how to run it locally to test,
so just submitted here to let the ci to test it.

roxell commented:
> > if you can add some tests would be great too
> 
> ok, I updated the test/test_lava_test_plans.py file with the lkft-android project, but I am not sure how to run it locally to test, so just submitted here to let the ci to test it.

make test
or a specific file:
pytest -v test/test_lava_job_visibility.py

roxell commented:
Think you have to update it like this http://ix.io/4i8d

roxell commented:
> Think you have to update it like this http://ix.io/4i8e

actually something like this http://ix.io/4i8g

roxell commented on test/test_lava_test_plans.py:
```diff
@@ -121,3 +121,25 @@ def test_call_lava_test_plan_testplans_ti(param):
+# lkft-android tests
+
+lkft_android_project_device_path = "lava_test_plans/projects/lkft-android/devices"
+lkft_android_project_testpaln_path = "lava_test_plans/projects/lkft-android/"
```
This isn't needed. please drop.

liuyq replied:
testpaln is necessary, there will be lkft-android specific testcases added there later.

roxell commented on test/test_lava_test_plans.py:
```diff
@@ -121,3 +121,25 @@ def test_call_lava_test_plan_testplans_ti(param):
+tests = []
+for device in lkft_android_devices:
+    for testcase in lkft_android_testcases:
+        tests.append((lkft_android_variable_input_file, device, testcase, lkft_android_project_device_path, lkft_android_project_testpaln_path))
```
drop here too

roxell commented on test/test_lava_test_plans.py:
```diff
@@ -121,3 +121,25 @@ def test_call_lava_test_plan_testplans_ti(param):
+@pytest.mark.parametrize("param", tests)
+def test_call_lava_test_plan_testcase_lkft_android(param):
+    variable_input_file, device, testcase, project_device_path, project_testplan_path= param
+    sys.argv = shlex.split(f'lava_test_plans --dry-run --variables "{variable_input_file}"  --testplan-device-path "{project_device_path}" --testplan-path "{project_testplan_path}" --device-type "{device}" --test-case "{testcase}" {test_lava_validity}')
```
the flag --testplan-path is used of you have a different path to the testplan's and in this test you don't use any plans you only use --test-case and not --test-plan, so please drop '--testplan-path "{project_testplan_path}"'

liuyq commented:
Hi, @roxell 

Updated according to your following review comments:

1. https://github.com/Linaro/lava-test-plans/pull/193/commits/e3d0cd1d91de9a2504c0792db69d31b34428ae14 
    should be squashed into the proper patch.

1. https://github.com/Linaro/lava-test-plans/pull/193/commits/b436152f07ddfe2e4bf4f80555506615292a737b
    Feels odd that we use people urls for a boot test job… not even sure if we need to add this switch or not. Maybe we should encourage people to update the postprocess step in the laba job file.. If people wants to update with custom things ?

1. https://github.com/Linaro/lava-test-plans/pull/193/commits/0c630ec21013505dc311d0f203895fb975a2d1a2
    please update the changelog –testplan-path shouldn’t be needed here.

Please help check it again to see if it's OK to be merged.

roxell commented on lava_test_plans/include/fastboot.jinja2:
```diff
@@ -22,7 +22,7 @@
 
 {# partitions for android jobs #}
-{% set partition_boot_a = partition_boot_a|default(false) %}
+{# boot_a partition would be set via boot_label #}
```
I assumed this should have been squashed into 220557ed56c2 ("include: fastboot: Add support for android partitions")

liuyq replied:
updated

roxell commented on lava_test_plans/include/fastboot.jinja2:
```diff
@@ -155,13 +155,7 @@ reboot_to_fastboot: {{ reboot_to_fastboot }}
-        compression: {{ BOOT_A_URL_COMP }}
-{% endif %}
-{% endif %}
+{# partition_boot_a will be done with the boot partition block #}
```
This too

liuyq replied:
updated

roxell commented on lava_test_plans/projects/lkft-android/include/metadata.jinja2:
```diff
@@ -0,0 +1,22 @@
+  android.build: "{{ BUILD_NUMBER }}"
+  android.name: "{{ JOB_NAME }}"
+  android.url: "{{ REFERENCE_BUILD_URL }}"
+  android.version: "{{ ANDROID_VERSION }}"
```
isn't "android." unnecessary ?

liuyq replied:
it's been used with the android prefix since the beginning, and I really think this makes it more clear to distinguish with others.

roxell replied:
ok

roxell commented on lava_test_plans/projects/lkft-android/include/metadata.jinja2:
```diff
@@ -0,0 +1,22 @@
+  android.name: "{{ JOB_NAME }}"
+  android.url: "{{ REFERENCE_BUILD_URL }}"
+  android.version: "{{ ANDROID_VERSION }}"
+{% if ANDROID_VENDOR_FINGERPRINT is defined %}
```
isn't 'ANDROID_'  prefix unnecessary ?

liuyq replied:
ANDROID_ makes it clear this field is on ANDROID instead of on other like, BUILD, LAVA, etc.

roxell replied:
ok

roxell commented on lava_test_plans/projects/lkft-android/include/metadata.jinja2:
```diff
@@ -0,0 +1,22 @@
+  git commit: "{{ KERNEL_COMMIT }}"
+  git describe: "{{ KERNEL_DESCRIBE }}"
+  build-url: "{{ BUILD_URL }}"
+  toolchain: "{{ TEST_METADATA_TOOLCHAIN }}"
```
isn't prefix 'TEST_METADATA_' unnecessary ?

liuyq replied:
TEST_METADATA means it's only used as TEST_METADATA, not build metadata, etc

roxell replied:
What do you mean with TEST_METADATA?

liuyq replied:
TEST means the variable is on lava test job, TEST_METADATA_ means the variable is used for the lava test job metadata section.
But here on this toolchain metatdata, as long as I know, it is not that useful for the Android failures investigation, and with the tuxsuite bake command it's impossible to get the toolchain information now, so I am thinking if it should be removed. what do you think about it?

roxell commented on lava_test_plans/include/test_target.jinja2:
```diff
@@ -1,5 +1,11 @@
+{% set USE_DOCKER_IMAGE_FOR_TEST_TARGET = use_docker_image_for_test_target|default(false) %}
```
should be in capital letters 'use_docker_image_for_test_target'.

liuyq commented on lava_test_plans/projects/lkft-android/include/metadata.jinja2:
```diff
@@ -0,0 +1,22 @@
+  git describe: "{{ KERNEL_DESCRIBE }}"
+  build-url: "{{ BUILD_URL }}"
+  toolchain: "{{ TEST_METADATA_TOOLCHAIN }}"
+  series: lkft
```
build-url is still used, the other 3 items could be deleted I think.

roxell commented:
you add this line:
{% set download_postprocess_required = download_postprocess_required|default(false) %}
in patch "lkft-android: add x15 and empty test boot job as reference"
and after that you add patch "include: fastboot: add flag for deploy download postprocess action" should be rearranged and added first.

liuyq commented:
> you add this line: {% set download_postprocess_required = download_postprocess_required|default(false) %} in patch "lkft-android: add x15 and empty test boot job as reference" and after that you add patch "include: fastboot: add flag for deploy download postprocess action" should be rearranged and added first.

Thanks, updated, please help check again.

roxell commented:
8512666b6d9f ("include: test_target: support to specify docker image") this patch includes two variables USE_DOCKER_IMAGE_FOR_TEST_TARGET and DOCKER_IMAGE_FOR_TEST they are included in a patch before, think they should be included in a patch after this patch instead.
or add the above patch before you include the variables in the variable.ini file.

liuyq commented:
> [8512666](https://github.com/Linaro/lava-test-plans/commit/8512666b6d9f5933078dd538328a75fb7f25fbf8) ("include: test_target: support to specify docker image") this patch includes two variables USE_DOCKER_IMAGE_FOR_TEST_TARGET and DOCKER_IMAGE_FOR_TEST they are included in a patch before, think they should be included in a patch after this patch instead. or add the above patch before you include the variables in the variable.ini file.

let me drop this commit from this PR, it's not used here since there is no real android test action added yet.
And I also moved the ("include: fastboot: add flag for deploy download postprocess action") commit earlier as you suggested before,  and removed more variables in the variables.ini file which are not used by this PR.

Please help check it again.
